funsor/tensor.py

```diff
@@ -0,0 +1,973 @@
+
+    # Compute inputs and outputs.
+    if lhs.inputs == rhs.inputs:
+        inputs, lhs_data, rhs_data = lhs.inputs, lhs.data, rhs.data
+    else:
+        inputs, (lhs_data, rhs_data) = align_tensors(lhs, rhs)
+    if len(lhs.output.shape) > 1:
+        rhs_data = rhs_data.reshape(rhs_data.shape + (1,) * (len(lhs.output.shape) - 1))
+
+    # Perform advanced indexing.
+    lhs_data_dim = len(lhs_data.shape)
+    target_dim = lhs_data_dim - len(lhs.output.shape) + op.offset
+    index = [None] * lhs_data_dim
+    for i in range(target_dim):
+        index[i] = ops.new_arange(lhs_data, lhs_data.shape[i]).reshape((-1,) + (1,) * (lhs_data_dim - i - 2))
+    index[target_dim] = rhs_data
+    for i in range(1 + target_dim, lhs_data_dim):
+        index[i] = ops.new_arange(lhs_data, lhs_data.shape[i]).reshape((-1,) + (1,) * (lhs_data_dim - i - 1))
+    data = lhs_data[tuple(index)]
+    return Tensor(data, inputs, lhs.dtype)
+
+
+@eager.register(Lambda, Variable, Tensor)
+def eager_lambda(var, expr):
+    inputs = expr.inputs.copy()
+    if var.name in inputs:
+        inputs.pop(var.name)
+        inputs[var.name] = var.output
+        data = align_tensor(inputs, expr)
+        inputs.pop(var.name)
+    else:
+        data = expr.data
+        shape = data.shape
+        dim = len(shape) - len(expr.output.shape)
+        data = data.reshape(shape[:dim] + (1,) + shape[dim:])
+        data = ops.expand(data, shape[:dim] + (var.dtype,) + shape[dim:])
+    return Tensor(data, inputs, expr.dtype)
+
+
+@dispatch(str, Variadic[Tensor])
+def eager_stack_homogeneous(name, *parts):
+    assert parts
+    output = parts[0].output
+    part_inputs = OrderedDict()
+    for part in parts:
+        assert part.output == output
+        assert name not in part.inputs
+        part_inputs.update(part.inputs)
+
+    shape = tuple(d.size for d in part_inputs.values()) + output.shape
+    data = ops.stack(0, *[ops.expand(align_tensor(part_inputs, part), shape)
+                          for part in parts])
+    inputs = OrderedDict([(name, bint(len(parts)))])
+    inputs.update(part_inputs)
+    return Tensor(data, inputs, dtype=output.dtype)
+
+
+@dispatch(str, str, Variadic[Tensor])
+def eager_cat_homogeneous(name, part_name, *parts):
+    assert parts
+    output = parts[0].output
+    inputs = OrderedDict([(part_name, None)])
+    for part in parts:
+        assert part.output == output
+        assert part_name in part.inputs
+        inputs.update(part.inputs)
+
+    tensors = []
+    for part in parts:
+        inputs[part_name] = part.inputs[part_name]
+        shape = tuple(d.size for d in inputs.values()) + output.shape
+        tensors.append(ops.expand(align_tensor(inputs, part), shape))
+    del inputs[part_name]
+
+    dim = 0
+    tensor = ops.cat(dim, *tensors)
+    inputs = OrderedDict([(name, bint(tensor.shape[dim]))] + list(inputs.items()))
+    return Tensor(tensor, inputs, dtype=output.dtype)
+
+
+def arange(prototype, name, *args, **kwargs):
```

Comment:
@fritzo As mentioned in a previous PR, I need `prototype` here to dispatch to `torch.arange` or `np.arange`. IMO, it is better to make `arange` and `materialize` Tensor methods, rather than standalone functions. WDYT?

Reply:
Sure, how about naming these methods `.new_arange()` and `.materialize()`?